operator/api/v1alpha2/seldondeployment_webhook.go

```diff
@@ -0,0 +1,446 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/seldonio/seldon-core/operator/constants"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"os"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"strconv"
+)
+
+// log is for logging in this package.
+var seldondeploymentlog = logf.Log.WithName("seldondeployment")
+
+func (r *SeldonDeployment) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+var _ webhook.Defaulter = &SeldonDeployment{}
+
+func GetContainerForPredictiveUnit(p *PredictorSpec, name string) *corev1.Container {
+	for j := 0; j < len(p.ComponentSpecs); j++ {
+		cSpec := p.ComponentSpecs[j]
+		for k := 0; k < len(cSpec.Spec.Containers); k++ {
+			c := &cSpec.Spec.Containers[k]
+			if c.Name == name {
+				return c
+			}
+		}
+	}
+	return nil
+}
+
+func GetPort(name string, ports []corev1.ContainerPort) *corev1.ContainerPort {
+	for i := 0; i < len(ports); i++ {
+		if ports[i].Name == name {
+			return &ports[i]
+		}
+	}
+	return nil
+}
+
+func IsPrepack(pu *PredictiveUnit) bool {
+	return *pu.Implementation == SKLEARN_SERVER || *pu.Implementation == XGBOOST_SERVER || *pu.Implementation == TENSORFLOW_SERVER || *pu.Implementation == MLFLOW_SERVER
+}
+
+func SetImageNameForPrepackContainer(pu *PredictiveUnit, c *corev1.Container) {
+	//Add missing fields
+	// Add image
+	if c.Image == "" {
+		if *pu.Implementation == SKLEARN_SERVER {
+
+			if pu.Endpoint.Type == REST {
+				c.Image = constants.DefaultSKLearnServerImageNameRest
+			} else {
+				c.Image = constants.DefaultSKLearnServerImageNameGrpc
+			}
+
+		} else if *pu.Implementation == XGBOOST_SERVER {
+
+			if pu.Endpoint.Type == REST {
+				c.Image = constants.DefaultXGBoostServerImageNameRest
+			} else {
+				c.Image = constants.DefaultXGBoostServerImageNameGrpc
+			}
+
+		} else if *pu.Implementation == TENSORFLOW_SERVER {
+
+			if pu.Endpoint.Type == REST {
+				c.Image = constants.DefaultTFServerImageNameRest
+			} else {
+				c.Image = constants.DefaultTFServerImageNameGrpc
+			}
+
+		} else if *pu.Implementation == MLFLOW_SERVER {
+
+			if pu.Endpoint.Type == REST {
+				c.Image = constants.DefaultMLFlowServerImageNameRest
+			} else {
+				c.Image = constants.DefaultMLFlowServerImageNameGrpc
+			}
+
+		}
+	}
+}
+
+// -----
+
+func addDefaultsToGraph(pu *PredictiveUnit) {
+	if pu.Type == nil {
+		ty := UNKNOWN_TYPE
+		pu.Type = &ty
+	}
+	if pu.Implementation == nil {
+		im := UNKNOWN_IMPLEMENTATION
+		pu.Implementation = &im
+	}
```

Comment:
Default pu.Endpoint.Type to "GRPC" since that is what SetImageNameForPrepackContainer() does?